actionpack/lib/action_controller/metal/strong_parameters.rb

```diff
@@ -284,10 +284,15 @@ def ==(other)
         end
       end
     end
-    alias eql? ==
+
+    def eql?(other)
+      self.class == other.class &&
+        permitted? == other.permitted? &&
+        parameters == other.parameters
+    end
 
     def hash
-      [@parameters.hash, @permitted].hash
+      [self.class, @parameters, @permitted].hash
```

Comment:
On reflection, I think `@permitted` should not be here, because it's mutable: we can distinguish it in `eql?`, but should not hash it.

Reply:
I'm unsure about this. In general in Ruby we do allow users to use mutable objects as Hash keys. It's just expected that they don't mutate them (or they call `#rehash`). `@parameters` is also mutable here, it's just less likely to be changed.

I think it may be best to keep this behaviour undefined, rather than hinting at a contract to users that we can't 100% fulfill.